### ISSUE TITLE
Add let operator

### ISSUE THREAD

**Issue by disnet** [mozilla/sweet.js#400]
We need a form of `operator` that does not bind the operator name in its template. Analogous to `let name = macro {}`.

```js
// stack overflow
operator + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
```

Some bikeshedding:

```js
// terrible, not analogous to real let
let operator + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
// ugly
let + = operator 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }

// or we could invert, current syntax doesn't bind in template
operator + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
// to opt into binding in the template
operator rec + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
rec operator + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
roperator + 12 left { $r, $l } => #{ (function(l, r) { return l + r})($r, $l0) }
```

Actually, it it ever necessary to have the operator bound in the template? Maybe we just drop any syntax that binds in the template?

**Comment by iffsid** [comment]
Any idea if this is/has been implemented (perhaps in a separate branch)?

The reason I ask is that for most cases, I can separate the use of the native operators into a separate module, import them, and redefine the operators using the `operator` macro:
``` js
var lops = require('./lifted-operators');

operator +   14 { $r } => #{ lops.add(0, $r) }
operator +   12 left { $l, $r } => #{ lops.add($l, $r) }
...
```
where, in `lifted-operators.js`, I can define the `add` function using `+`.

However, there is a particular use case for operator macros, where it is desirable for the redefined operators to have _global_ scope when necessary. For example, when doing [abstract interpretation](https://en.wikipedia.org/wiki/Abstract_interpretation), which in turn, typically requires operator overloading.
The idea there is to take whatever js program desired, with all the respective `require` statements, and have the redefined operators take global scope for the program *and* its dependencies.

With this in mind, I have two (inter-related) questions:
1. What would be the most effective way of achieving the requirement for *global* operator rewriting?
2. If I have a separate `operators.sjs` file with the redefined macros, how could I export the redefined operators without naming the exports differently?

**Comment by disnet** [comment]
> Any idea if this is/has been implemented (perhaps in a separate branch)?

No I haven't had time to work on this yet.

> 1. What would be the most effective way of achieving the requirement for global operator rewriting?

You mean have `sjs` automatically walk each `require` statement and install the custom `+` etc. operators? Currently the only hacky way I can think of would be to write a `require` macro that uses the path to the dependency and compiles it to a new file using the sweet.js API with your custom operators as a module and then expand to a normal require with its path pointing to the freshly compiled file.

We should be able to make this less hacky when es6-modules lands.

> 2. If I have a separate operators.sjs file with the redefined macros, how could I export the redefined operators without naming the exports differently?

Sorry, I'm not sure what you're asking.

**Comment by iffsid** [comment]
> You mean have sjs automatically walk each require statement and install the custom + etc. operators? Currently the only hacky way I can think of would be to write a require macro that uses the path to the dependency and compiles it to a new file using the sweet.js API with your custom operators as a module and then expand to a normal require with its path pointing to the freshly compiled file.

Right. That's what I figured I'd have to do.
I was planning on using a version of `sweetify` with `browserify` that didn't restrict itself to `.sjs` files.

> We should be able to make this less hacky when es6-modules lands.

Ah I see. Thanks.

>> 2. If I have a separate operators.sjs file with the redefined macros, how could I export the redefined operators without naming the exports differently?
> Sorry, I'm not sure what you're asking.

Apologies, I'm not entirely sure how to frame the question.
Suppose I had a macros file `operators.sjs` that looked like this:
``` js
operator +   14 { $r } => #{ ad.add(0, $r) }
...
```
How would/should I write the exports and/or requires such that I can use these in a node-loader style to do what I wanted in the other question; basically have these operator-macros apply to those other files. I'm thinking this will require, in say `foo.js` where I want to apply the macros, something like:
``` js
var sjs = require('sweet.js');
var ops = require('./macros/operators.sjs')
```
but I don't quite understand
a. how to export operator macros -- I don't seem to be able to write
``` js
module.exports = {
  +: +,
  ...
}
```
b. how not to then have to refer to the macros themselves as ops.+ within `foo.js`

I'm not sure if that explains the question any better. :-)

**Comment by disnet** [comment]
Ah! Got it I think.

Yeah, so this one gets tricky. The "right" answer is wait for es6-modules but since I'm in the middle of thesis writing this might be a while.

But there are still some hacky solutions in the meantime.

Fist, to your question about exporting operators, it's just like exporting normal macros in sweet.js so to export `+` you just `export +`. But we currently have no way of namespaceing them on import, so you can't do `ops.+` (in es6-modules you'll be able to though!).

The simplest solution for you might be a harness snippet you just concatenate with the files you want to compile. So `harness.js` might look like:

```js
var lops = require("lifted-operators");
operator + 14 { $r } => #{ lops.add(0, $r) }
// other custom operators
```

Then you can just add it to the beginning of every file you want to compile:


```js
var harness = fs.readFileSync("harness.js");
sweet.compile(harness + foo)
// ...
```

Adding the harness and compiling could be done either through your deep `require` macro or if you don't need to automatically walk the require paths you could just do this as a little `sjs` replacement compile script. This is basically what I did with [sweet-virtual-values](https://github.com/disnet/sweet-virtual-values).

Does that help?

**Comment by iffsid** [comment]
Ah I see. Ok I think I understand better now.
The issue was I *didn't* want to refer the the operators as `ops.+`. Basically, there are these preexisting files that use the various js operators `+`, `-`, etc. and I want to inject my lifted functions into those locations; e.g. turn `a+b` into `ops.mul(a,b)` using sweetjs macros, by making the most minimal of changes to the prexisting files, ideally none at all, but I don't know if that's likely.

What I have now is basically the same as what you suggest, except I abstracted the `require` statement in `harness.js` into a macro of its own.
I did this mainly because I wanted to be able to run both on the command-line and on the browser.
So now, to run in the browser, I can call `browserify` with the `sweetify` transformation that should compile things appropriately with the macros.
For the command-line use, I just do what you suggest and call `sweet.compile` with the requisite `require` and appropriate macros and things should work.

I'm quite pleased with this setup. This gives me a quite general setup to do abstract interpretation in javascript.
I'm using all this for the [ad](https://github.com/iffsid/ad.js) package, which should allow for the computation of the *gradients* of arbitrary programs in javascript ([More](https://en.wikipedia.org/wiki/Automatic_differentiation) on AD).
Good luck with the thesis!
